### PR TITLE
pleroma: init at 2.3.50

### PR DESCRIPTION
**Pull request by ghost**
###### Motivation for this change
Thanks a lot to @happysalada for restoring buildMix and integrating it with mixRelease! :slightly_smiling_face: 
A variant of this is deployed on https://pleroma.pbb.lc
This can be used with the original module from https://github.com/NixOS/nixpkgs/pull/103138

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Comment by happysalada** [comment]
Some small nits
- How about running nixpkgs-fmt on mix.nix ? I know the defaullt format of mix.nix uses tabs, but just so we agree.
- How about using substituteInPlace instead of sed ? I find it a bit clearer, but I know it's a matter of preference so no worries if you would rather use sed.

**Comment by happysalada** [comment]
Result of `nixpkgs-review pr 125934` run on x86_64-darwin [1](https://github.com/Mic92/nixpkgs-review)

**Comment by happysalada** [comment]
Two brief updates.
The mistaken jose identified issue has been fixed on mix2nix master.
I just opened an issue upstream about the formatting issue with nixpkgs
https://github.com/ydlr/mix2nix/issues/8

**Comment by happysalada** [comment]
I've merged a PR to reduce the size of the resulting mix-release closure in https://github.com/NixOS/nixpkgs/pull/125885
If you have a moment to rebase against master, I would be curious to test the effect on this PR.

File: pkgs/servers/pleroma/default.nix [hunk 1]
```diff
@@ -0,0 +1,206 @@
...
+        version = "1.0.18";
+
+        src = beamPackages.fetchHex {
+          pkg = "${name}";
+          version = "${version}";
+          sha256 = "0fnx2pm1n2m0zs2skivv43s42hrgpq9i143p9mngw9f3swjqpxvx";
+        };
+
+        patchPhase = ''
+          echo '{plugins, [ { pc, { git, "git://github.com/blt/port_compiler.git", {tag, "v${beamPackages.pc.version}"}}} ]}.' >> rebar.config
```
Comment (happysalada) [inline pkgs/servers/pleroma/default.nix:198]: @dlesl just in case you have something to add on this

**Comment by happysalada** [comment]
A new version of mix2nix has been published with the formatting issue and the jose issue fixed.
The update has been merged to nixpkgs.
Just in case you want to use it.

**Comment by happysalada** [comment]
One more thing. How about removing pleroma-otp (which is the current pleroma package downloaded basically).
And defining the default pleroma package of the service as this one?

**Comment by petabyteboy** [comment]
I have implemented all the suggested changes :slightly_smiling_face: 
The new mix2nix correctly identified `jose`, but incorrectly identified `base64url` :see_no_evil: 
The closure size was greatly reduced from ~900M to ~270M.

File: pkgs/servers/pleroma/default.nix (hunk 1 above)
Comment (dlesl) [reply pkgs/servers/pleroma/default.nix:198]: It looks like this library uses `rebar`, not `rebar3`, so I think injecting this here is a reasonable thing to do. However it is not necessary to specify a version, `{plugins, [pc]}` should do -- `rebar3WithPlugins` will not try to fetch it from `hex.pm`

**Comment by happysalada** [comment]
Just for reference, I've created an issue upstream in mix2nix https://github.com/ydlr/mix2nix/issues/9

File: pkgs/development/beam-modules/build-mix.nix [hunk 2]
```diff
@@ -23,7 +24,7 @@ let
 
   pkg = self: stdenv.mkDerivation (attrs // {
     name = "${name}-${version}";
-    inherit version src buildInputs;
+    inherit version src;
```
Comment (happysalada) [inline pkgs/development/beam-modules/build-mix.nix:27]: just putting it here so we don't forget.
Those changes were merged in a separate PR so you'll probably need to rebase

**Comment by happysalada** [comment]
Also very random question. Did you try running that already for a little bit?
I had some weird error on app I'm running using Tesla as well when I upgraded to erlang24. Some http request failing with the `:nxdomain` error. Wondering if you saw the same. If you did not, no worries.

**Comment by petabyteboy** [comment]
> Also very random question. Did you try running that already for a little bit?
> I had some weird error on app I'm running using Tesla as well when I upgraded to erlang24. Some http request failing with the `:nxdomain` error. Wondering if you saw the same. If you did not, no worries.

I have been running it for a bit now, and haven't noticed anything strange with it, but I also didn't take a closer look at the log yet.

**Comment by NinjaTrappeur** [comment]
NICE!!

File: pkgs/development/beam-modules/build-mix.nix (hunk 2 above)
Comment (petabyteboy) [reply pkgs/development/beam-modules/build-mix.nix]: done

**Comment by happysalada** [comment]
I have no more nitpicks.
Perhaps david's comment about not specifying the version.

Personally I'm fine to merge anytime. Feel free to change the status to not draft and ping me when you want to merge.

**Comment by petabyteboy** [comment]
I should probably add some meta attributes and stuff. @NinjaTrappeur do you want to be a maintainer for this package as well?

File: pkgs/servers/pleroma/default.nix (hunk 1 above)
Comment (petabyteboy) [reply pkgs/servers/pleroma/default.nix:198]: Thanks for the hint, will implement

**Comment by NinjaTrappeur** [comment]
> I should probably add some meta attributes and stuff. @NinjaTrappeur do you want to be a maintainer for this package as well?

Sure!

I don't have much experience with mix.nix (nor a particular interest in
it) though. Don't expect much from me appart reviews and occasional
version bumps.

**Comment by happysalada** [comment]
@NinjaTrappeur version bumps might not be completely trivial (might need to modify git dependencies manually), but if you experience a problem, feel free to ping me and I can have a look.

**Comment by NinjaTrappeur** [comment]
The PR consistenly (tried 5 times) triggers a kernel panic when running the VM test:

```
pleroma # [   49.100564] pleroma[976]: 13:41:29.092 [error] The :mime library has been compiled with the following custom types:
pleroma # [   49.101578] pleroma[976]:     %{}
pleroma # [   49.101944] pleroma[976]: But it is being started with the following types:
pleroma # [   49.102576] pleroma[976]:     %{"application/activity+json" => ["activity+json"], "application/jrd+json" => ["jrd+json"], "application/ld+json" => ["activity+json"], "application/xml" => ["xml"], "application/xrd+xml" => ["xrd+xml"]}
pleroma # [   49.104293] pleroma[976]: We are going to dynamically recompile it during boot,
pleroma # [   49.104959] pleroma[976]: but please clean the :mime dependency to make sure it is recompiled:
pleroma # [   49.108149] pleroma[976]:     $ mix deps.clean mime --build
pleroma # [   49.108998] pleroma[976]: warning: redefining module MIME (current version loaded from /nix/store/64hv4rk1fzgiqahvdjxp1xinjh3y6vbp-pleroma-2.3.50/lib/mime-1.6.0/ebin/Elixir.MIME.beam)
pleroma # [   49.114473] pleroma[976]:   /build/hex-source-mime-1.6.0/lib/mime/application.ex:27
pleroma # [   49.947130] 1_scheduler invoked oom-killer: gfp_mask=0x100dca(GFP_HIGHUSER_MOVABLE|__GFP_ZERO), order=0, oom_score_adj=0
pleroma # [   49.948306] CPU: 0 PID: 1023 Comm: 1_scheduler Not tainted 5.10.40 #1-NixOS
pleroma # [   49.949048] Hardware name: QEMU Standard PC (i440FX + PIIX, 1996), BIOS rel-1.14.0-0-g155821a1990b-prebuilt.qemu.org 04/01/2014
pleroma # [   49.950224] Call Trace:
pleroma # [   49.950517]  dump_stack+0x6b/0x83
pleroma # [   49.950872]  dump_header+0x4a/0x1ed
pleroma # [   49.951243]  out_of_memory.cold+0xa/0x7e
pleroma # [   49.951667]  __alloc_pages_slowpath.constprop.0+0xbd1/0xca0
pleroma # [   49.952246]  __alloc_pages_nodemask+0x308/0x340
pleroma # [   49.952727]  alloc_pages_vma+0x80/0x250
pleroma # [   49.953133]  handle_mm_fault+0xe43/0x16c0
pleroma # [   49.953564]  do_user_addr_fault+0x1bb/0x3f0
pleroma # [   49.954004]  exc_page_fault+0x5d/0x120
pleroma # [   49.954400]  ? asm_exc_page_fault+0x8/0x30
pleroma # [   49.954866]  asm_exc_page_fault+0x1e/0x30
pleroma # [   49.955261] RIP: 0033:0x677359
pleroma # [   49.955571] Code: 01 00 00 48 83 fa 03 0f 86 c4 01 00 00 4c 89 d3 ba 08 00 00 00 31 f6 48 d1 eb 0f 1f 80 00 00 00 00 f3 0f 6f 0c 11 48 83 c6 01 <0f> 11 0c 10 48 83 c2 10 48 39 f3 75 ea 4c 89 d6 48 83 e6 fe 48 8d
pleroma # [   49.957366] RSP: 002b:00007fa8c63229c0 EFLAGS: 00010206
pleroma # [   49.957876] RAX: 00007fa88d62efd8 RBX: 0000000000000003 RCX: 00007fa88d62eb90
pleroma # [   49.958568] RDX: 0000000000000028 RSI: 0000000000000003 RDI: 00007fa88d62efe0
pleroma # [   49.959255] RBP: 00007fa8c6322a40 R08: 00007fa88d62efc9 R09: 00007fa88d62eb98
pleroma # [   49.959962] R10: 0000000000000007 R11: 0000000000000006 R12: 00007fa8c6322a80
pleroma # [   49.960653] R13: 0000000000000000 R14: 0000000000000001 R15: 0000000000000001
pleroma # [   49.961366] Mem-Info:
pleroma # [   49.961602] active_anon:2702 inactive_anon:71248 isolated_anon:0
pleroma # [   49.961602]  active_file:109 inactive_file:120 isolated_file:0
pleroma # [   49.961602]  unevictable:0 dirty:0 writeback:0
pleroma # [   49.961602]  slab_reclaimable:3754 slab_unreclaimable:3883
pleroma # [   49.961602]  mapped:14187 shmem:16741 pagetables:637 bounce:0
pleroma # [   49.961602]  free:1169 free_pcp:0 free_cma:0
pleroma # [   49.964402] Node 0 active_anon:10808kB inactive_anon:284992kB active_file:436kB inactive_file:480kB unevictable:0kB isolated(anon):0kB isolated(file):0kB mapped:56748kB dirty:0kB writeback:0kB shmem:66964kB shmem_thp: 0kB shmem_pmdmapped: 0kB anon_thp: 0kB writeback_tmp:0kB kernel_stack:1696kB all_unreclaimable? no
pleroma # [   49.966899] Node 0 DMA free:1400kB min:104kB low:128kB high:152kB reserved_highatomic:0KB active_anon:1144kB inactive_anon:13040kB active_file:0kB inactive_file:0kB unevictable:0kB writepending:0kB present:15992kB managed:15908kB mlocked:0kB pagetables:40kB bounce:0kB free_pcp:0kB local_pcp:0kB free_cma:0kB
pleroma # [   49.969321] lowmem_reserve[]: 0 326 326 326 326
pleroma # [   49.969736] Node 0 DMA32 free:3276kB min:7332kB low:7896kB high:8460kB reserved_highatomic:0KB active_anon:9664kB inactive_anon:271932kB active_file:436kB inactive_file:472kB unevictable:0kB writepending:0kB present:376684kB managed:348136kB mlocked:0kB pagetables:2508kB bounce:0kB free_pcp:0kB local_pcp:0kB free_cma:0kB
pleroma # [   49.972295] lowmem_reserve[]: 0 0 0 0 0
pleroma # [   49.972651] Node 0 DMA: 20*4kB (U) 23*8kB (UM) 9*16kB (UM) 5*32kB (UM) 3*64kB (U) 3*128kB (U) 1*256kB (U) 0*512kB 0*1024kB 0*2048kB 0*4096kB = 1400kB
pleroma # [   49.973866] Node 0 DMA32: 34*4kB (M) 34*8kB (UM) 116*16kB (UME) 32*32kB (UE) 0*64kB 0*128kB 0*256kB 0*512kB 0*1024kB 0*2048kB 0*4096kB = 3288kB
pleroma # [   49.975036] Node 0 hugepages_total=0 hugepages_free=0 hugepages_surp=0 hugepages_size=1048576kB
pleroma # [   49.975816] Node 0 hugepages_total=0 hugepages_free=0 hugepages_surp=0 hugepages_size=2048kB
pleroma # [   49.976604] 16981 total pagecache pages
pleroma # [   49.976969] 0 pages in swap cache
pleroma # [   49.977273] Swap cache stats: add 0, delete 0, find 0/0
pleroma # [   49.977750] Free swap  = 0kB
pleroma # [   49.978030] Total swap = 0kB
pleroma # [   49.978295] 98169 pages RAM
pleroma # [   49.978555] 0 pages HighMem/MovableOnly
pleroma # [   49.978906] 7158 pages reserved
pleroma # [   49.979212] 0 pages cma reserved
pleroma # [   49.979511] Tasks state (memory values in pages):
pleroma # [   49.979954] [  pid  ]   uid  tgid total_vm      rss pgtables_bytes swapents oom_score_adj name
pleroma # [   49.980733] [    459]     0   459     5904      167    69632        0          -250 systemd-journal
pleroma # [   49.981565] [    475]     0   475     5491      247    61440        0         -1000 systemd-udevd
pleroma # [   49.982396] [    574]     0   574    56129      209    65536        0             0 sh
pleroma # [   49.983124] [    602]     0   602     2286      114    61440        0             0 systemd-logind
pleroma # [   49.983944] [    656]     4   656     1630      109    53248        0          -900 dbus-daemon
pleroma # [   49.984723] [    731] 64015   731   233177       74   147456        0             0 nscd
pleroma # [   49.985463] [    746]     0   746      797      102    45056        0             0 dhcpcd
pleroma # [   49.986214] [    792]     0   792    55630       10    53248        0             0 agetty
pleroma # [   49.986972] [    797]    60   797     5694      304    77824        0             0 nginx
pleroma # [   49.987720] [    800]    60   800     5836      418    81920        0             0 nginx
pleroma # [   49.988483] [    822]    71   822    73059     1426   135168        0             0 postgres
pleroma # [   49.989248] [    824]    71   824    73091      601   122880        0             0 postgres
pleroma # [   49.990018] [    825]    71   825    73059      291   122880        0             0 postgres
pleroma # [   49.990775] [    826]    71   826    73059      434   114688        0             0 postgres
pleroma # [   49.991545] [    827]    71   827    73198      532   122880        0             0 postgres
pleroma # [   49.992312] [    828]    71   828    69236      293   102400        0             0 postgres
pleroma # [   49.993097] [    829]    71   829    73164      410   118784        0             0 postgres
pleroma # [   49.993856] [    976]   999   976   548839    78258   864256        0             0 beam.smp
pleroma # [   49.994626] [   1014]   999  1014     1624       59    53248        0             0 epmd
pleroma # [   49.995477] [   1017]     0  1017    56129      208    57344        0             0 sh
pleroma # [   49.996526] [   1019]   999  1019      577       10    40960        0             0 erl_child_setup
pleroma # [   49.997668] [   1020]     0  1020    56133      208    57344        0             0 sh
pleroma # [   49.998696] [   1021]     0  1021    56796       73    65536        0             0 timeout
pleroma # [   49.999660] [   1022]     0  1022    56106      144    61440        0             0 bash
pleroma # [   50.000400] [   1055]     0  1055    56796       74    65536        0             0 sleep
pleroma # [   50.001159] Kernel panic - not syncing: Out of memory: compulsory panic_on_oom is enabled
pleroma # [   50.001903] CPU: 0 PID: 1023 Comm: 1_scheduler Not tainted 5.10.40 #1-NixOS
pleroma # [   50.002535] Hardware name: QEMU Standard PC (i440FX + PIIX, 1996), BIOS rel-1.14.0-0-g155821a1990b-prebuilt.qemu.org 04/01/2014
pleroma # [   50.003575] Call Trace:
pleroma # [   50.003807]  dump_stack+0x6b/0x83
pleroma # [   50.004109]  panic+0x101/0x2c8
pleroma # [   50.004389]  out_of_memory.cold+0x2f/0x7e
pleroma # [   50.004756]  __alloc_pages_slowpath.constprop.0+0xbd1/0xca0
pleroma # [   50.005304]  __alloc_pages_nodemask+0x308/0x340
pleroma # [   50.005749]  alloc_pages_vma+0x80/0x250
pleroma # [   50.006124]  handle_mm_fault+0xe43/0x16c0
pleroma # [   50.006522]  do_user_addr_fault+0x1bb/0x3f0
pleroma # [   50.006931]  exc_page_fault+0x5d/0x120
pleroma # [   50.007300]  ? asm_exc_page_fault+0x8/0x30
pleroma # [   50.007688]  asm_exc_page_fault+0x1e/0x30
pleroma # [   50.008047] RIP: 0033:0x677359
pleroma # [   50.008327] Code: 01 00 00 48 83 fa 03 0f 86 c4 01 00 00 4c 89 d3 ba 08 00 00 00 31 f6 48 d1 eb 0f 1f 80 00 00 00 00 f3 0f 6f 0c 11 48 83 c6 01 <0f> 11 0c 10 48 83 c2 10 48 39 f3 75 ea 4c 89 d6 48 83 e6 fe 48 8d
pleroma # [   50.010013] RSP: 002b:00007fa8c63229c0 EFLAGS: 00010206
pleroma # [   50.010489] RAX: 00007fa88d62efd8 RBX: 0000000000000003 RCX: 00007fa88d62eb90
pleroma # [   50.011131] RDX: 0000000000000028 RSI: 0000000000000003 RDI: 00007fa88d62efe0
pleroma # [   50.011769] RBP: 00007fa8c6322a40 R08: 00007fa88d62efc9 R09: 00007fa88d62eb98
pleroma # [   50.012405] R10: 0000000000000007 R11: 0000000000000006 R12: 00007fa8c6322a80
pleroma # [   50.013045] R13: 0000000000000000 R14: 0000000000000001 R15: 0000000000000001
pleroma # [   50.013808] Kernel Offset: 0x7a00000 from 0xffffffff81000000 (relocation range: 0xffffffff80000000-0xffffffffbfffffff)
pleroma # [   50.014791] Rebooting in 1 seconds..
```

I consistently (tried 3 times) do not get this error on the current master branch. I'll have a further look later on.

**Comment by petabyteboy** [comment]
Whut... My pleroma is using 161M RES.

**Comment by NinjaTrappeur** [comment]
I just had a further look. 

I suspect the mime extension to be the source of the OOM. It seems like we lack some `MIME type -> extension` configuration at build time, forcing us to re-compile the library on runtime. The compilation then OOM the VM test.

```
pleroma # [   45.980735] pleroma[981]: 19:04:24.246 [error] The :mime library has been compiled with the following custom types:
pleroma # [   45.981778] pleroma[981]:     %{}
pleroma # [   45.982437] pleroma[981]: But it is being started with the following types:
pleroma # [   45.983391] pleroma[981]:     %{"application/activity+json" => ["activity+json"], "application/jrd+json" => ["jrd+json"], "application/ld+json" => ["activity+json"], "application/xml" => ["xml"], "application/xrd+xml" => ["xrd+xml"]}
pleroma # [   45.980735] pleroma[981]: 19:04:24.246 [error] The :mime library has been compiled with the following custom types:
pleroma # [   45.981778] pleroma[981]:     %{}
pleroma # [   45.982437] pleroma[981]: But it is being started with the following types:
pleroma # [   45.983391] pleroma[981]:     %{"application/activity+json" => ["activity+json"], "application/jrd+json" => ["jrd+json"], "application/ld+json" => ["activity+json"], "application/xml" => ["xml"], "application/xrd+xml" => ["xrd+xml"]}
pleroma # [   45.985446] pleroma[981]: We are going to dynamically recompile it during boot,
pleroma # [   45.987020] pleroma[981]: but please clean the :mime dependency to make sure it is recompiled:
pleroma # [   45.989796] pleroma[981]:     $ mix deps.clean mime --build
pleroma # [   45.990355] pleroma[981]: warning: redefining module MIME (current version loaded from /nix/store/64hv4rk1fzgiqahvdjxp1xinjh3y6vbp-pleroma-2.3.50/lib/mime-1.6.0/ebin/Elixir.MIME.beam)
pleroma # [   45.994093] pleroma[981]:   /build/hex-source-mime-1.6.0/lib/mime/application.ex:27
pleroma # [   48.835501] 1_dirty_cpu_sch invoked oom-killer: gfp_mask=0x100cca(GFP_HIGHUSER_MOVABLE), order=0, oom_score_adj=0
pleroma # [   48.836438] CPU: 0 PID: 1031 Comm: 1_dirty_cpu_sch Not tainted 5.10.40 #1-NixOS
pleroma # [   48.837077] Hardware name: QEMU Standard PC (i440FX + PIIX, 1996), BIOS rel-1.14.0-0-g155821a1990b-prebuilt.qemu.org 04/01/2014
pleroma # [   48.838084] Call Trace:

(...)
```

How could we add this `%{"application/activity+json" => ["activity+json"], "application/jrd+json" => ["jrd+json"], "application/ld+json" => ["activity+json"], "application/xml" => ["xml"], application/xrd+xml" => ["xrd+xml"]}` config to the mime package?

**Comment by petabyteboy** [comment]
I've been ignoring this warning ever since I started using pleroma, but sure, we can try

**Comment by NinjaTrappeur** [comment]
I'm not 100% sure this is the root cause of the issue. That's the only lead I have for now though.

After a few tests, it seems like we're OOMing with 256MB but are doing fine with 512. Another way to go if we're stuck with this OOM issue would be to increase the test VM memory through `virtualisation.memorySize = 512;`.

**Comment by petabyteboy** [comment]
@NinjaTrappeur you were completely right, this fixes the issue. Thanks a lot!

**Comment by petabyteboy** [comment]
Please note that this is not a stable release. The latest stable release, 2.3.0, does not build with Erlang/OTP 24 for various reasons. We can either merge this non-release version or wait for the next release to happen. I'm running this on my instance either way.
I haven't found a good way to set the Erlang/OTP version, but it will likely be required to change pleroma to a fixed Erlang/OTP version when we update the default one.

**Comment by happysalada** [comment]
I don't think that it being not a stable release is a blocker.

File: pkgs/top-level/all-packages.nix [hunk 3]
```diff
@@ -7867,7 +7867,7 @@ in
 
   tautulli = python3Packages.callPackage ../servers/tautulli { };
 
-  pleroma-otp = callPackage ../servers/pleroma-otp { };
+  pleroma = callPackage ../servers/pleroma { };
```
Comment (SuperSandro2000) [inline pkgs/top-level/all-packages.nix:7870]: We should add an entry to aliases.nix

Comment (petabyteboy) [reply pkgs/top-level/all-packages.nix:7870]: It's not really the same package and I doubt anyone used the package directly, but sure, adding it won't hurt either

Comment (petabyteboy) [reply pkgs/top-level/all-packages.nix:7870]: done

**Comment by happysalada** [comment]
Thanks again for this PR!

**Pull request merged**